### PR TITLE
Bluetooth: Shell: Make `gatt notify` more general

### PR DESCRIPTION
**Pull request by alwa-nordic**
Make `gatt notify` take the source handle and the data to send.

### CODE REVIEW

File: subsys/bluetooth/shell/gatt.c [hunk 1]
```diff
@@ -1294,7 +1328,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(gatt_cmds,
 	SHELL_CMD_ARG(unregister, NULL,
 		      "unregister pre-predefined test service",
 		      cmd_unregister_test_svc, 1, 0),
-	SHELL_CMD_ARG(notify, NULL, "[data]", cmd_notify, 1, 1),
+	SHELL_CMD_ARG(notify, NULL, "<handle> <conn> <str>", cmd_notify, 3, 0),
```
Comment (theob-pro) [inline subsys/bluetooth/shell/gatt.c:1331]: The number of mandatory arguments should be 4 because it include the command name.

Comment (alwa-nordic) [reply subsys/bluetooth/shell/gatt.c:1331]:
```suggestion
	SHELL_CMD_ARG(notify, NULL, "<handle> <conn> <str>", cmd_notify, 4, 0),
```

File: subsys/bluetooth/shell/gatt.c [hunk 2]
```diff
@@ -1294,7 +1328,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(gatt_cmds,
 	SHELL_CMD_ARG(unregister, NULL,
 		      "unregister pre-predefined test service",
 		      cmd_unregister_test_svc, 1, 0),
-	SHELL_CMD_ARG(notify, NULL, "[data]", cmd_notify, 1, 1),
+	SHELL_CMD_ARG(notify, NULL, "<handle> <conn> <str>", cmd_notify, 4, 0),
```
Comment (alwa-nordic) [inline subsys/bluetooth/shell/gatt.c:1331]:
```suggestion
	SHELL_CMD_ARG(notify, NULL, "<handle> <str>", cmd_notify, 3, 0),
```

File: subsys/bluetooth/shell/gatt.c [hunk 3]
```diff
@@ -1294,7 +1328,7 @@ SHELL_STATIC_SUBCMD_SET_CREATE(gatt_cmds,
 	SHELL_CMD_ARG(unregister, NULL,
 		      "unregister pre-predefined test service",
 		      cmd_unregister_test_svc, 1, 0),
-	SHELL_CMD_ARG(notify, NULL, "[data]", cmd_notify, 1, 1),
+	SHELL_CMD_ARG(notify, NULL, "<handle> <str>", cmd_notify, 3, 0),
```
Comment (jori-nordic) [inline subsys/bluetooth/shell/gatt.c:1331]: could we have an arg for different data types:
e.g. `gatt notify [handle] bin deadbeef1337` will write those hex values and `gatt notify [handle] str "hello world"` will write the "hello world" string.

Comment (jori-nordic) [reply subsys/bluetooth/shell/gatt.c:1331]: or make two commands: notify-str and notify-bin. But I think it's gonna duplicate code a bit

Comment (alwa-nordic) [reply subsys/bluetooth/shell/gatt.c:1331]: I would prefer if the shell handled this somehow. Maybe like \x"deadbeef1337"?

File: subsys/bluetooth/shell/gatt.c [hunk 4]
```diff
@@ -946,43 +949,73 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
+	return BT_GATT_ITER_STOP;
+}
+
+static const struct bt_gatt_attr *find_attr(uint16_t handle)
+{
+	const struct bt_gatt_attr *attr = NULL;
+
+	bt_gatt_foreach_attr(handle, handle, found_attr, &attr);
+
+	return attr;
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:968]: Should we consider adding a `bt_gatt_find_by_handle` function?

File: subsys/bluetooth/shell/gatt.c [hunk 5]
```diff
@@ -946,43 +949,73 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
 
-	if (argc > 1) {
-		data = strtoul(argv[1], NULL, 16);
+	err = 0;
+	handle = shell_strtoul(arg_handle, 16, &err);
+	if (err) {
+		shell_error(sh, "Handle: Not a hex number '%s'.", arg_handle);
+		return -EINVAL;
+	}
+	if (!IN_RANGE(handle, 1, 0xffff)) {
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:987]:
```suggestion
	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
```

File: subsys/bluetooth/shell/gatt.c [hunk 6]
```diff
@@ -946,43 +949,73 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
-	params.data = &data;
-	params.len = sizeof(data);
-	params.func = notify_cb;
-	params.user_data = (void *)sh;
-	SET_CHAN_OPT_ANY(params);
+	data_len = strnlen(arg_data, UINT16_MAX + 1);
+	if (data_len > UINT16_MAX) {
+		shell_error(sh, "The argument is too big to fit in a notification.");
+		return -EINVAL;
+	}
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:1002]: Should we consider using `hex2bin` here instead? Or is your intention to only support string values?

File: subsys/bluetooth/shell/gatt.c (hunk 3 above)
Comment (Thalley) [reply subsys/bluetooth/shell/gatt.c:1330]: Sending hex or string will work the same for us if using `hex2bin`. The receiving side can then choose whether or not to display the data as a string or hex.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/shell/gatt.c (hunk 4 above)
Comment (alwa-nordic) [reply subsys/bluetooth/shell/gatt.c:968]: Yes, I think so. It's already implemented as the internal function `find_attr` in `gatt.c`.

File: subsys/bluetooth/shell/gatt.c (hunk 3 above)
Comment (alwa-nordic) [reply subsys/bluetooth/shell/gatt.c]: Updated PR: The command now takes a hex value, like other commands like `gatt write`.

File: subsys/bluetooth/shell/gatt.c [hunk 7]
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
+		shell_error(sh, "Handle '%s': Not a valid hex number.", arg_handle);
+		return -EINVAL;
+	}
+	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
+		shell_error(sh, "Handle 0x%lx: Impossible value.", handle);
+		return -EINVAL;
 	}
 
-	memset(&params, 0, sizeof(params));
+	if ((arg_data_len/2) > CHAR_SIZE_MAX) {
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:995]:
```suggestion
	if ((arg_data_len / 2) > CHAR_SIZE_MAX) {
```

File: subsys/bluetooth/shell/gatt.c [hunk 8]
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
+	size_t data_len;
+	unsigned long handle;
+	static char data[CHAR_SIZE_MAX];
 
-	if (!echo_enabled) {
-		shell_error(sh, "No clients have enabled notifications for the vnd1_echo CCC.");
-		return -ENOEXEC;
-	}
+	char *arg_handle = argv[1];
+	char *arg_data = argv[2];
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:981]:
```suggestion
	const char *arg_handle = argv[1];
	const char *arg_data = argv[2];
```

File: subsys/bluetooth/shell/gatt.c [hunk 9]
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
 
-	if (argc > 1) {
-		data = strtoul(argv[1], NULL, 16);
+	err = 0;
+	handle = shell_strtoul(arg_handle, 16, &err);
+	if (err) {
+		shell_error(sh, "Handle '%s': Not a valid hex number.", arg_handle);
+		return -EINVAL;
+	}
+	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:990]:
```suggestion
	}
	
	if (!IN_RANGE(handle, BT_ATT_FIRST_ATTRIBUTE_HANDLE, BT_ATT_LAST_ATTRIBUTE_HANDLE)) {
```

File: subsys/bluetooth/shell/gatt.c [hunk 10]
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
+		shell_error(sh, "Handle 0x%lx: Impossible value.", handle);
+		return -EINVAL;
 	}
 
-	memset(&params, 0, sizeof(params));
+	if ((arg_data_len/2) > CHAR_SIZE_MAX) {
+		shell_error(sh, "Data: Size exceeds legal attribute size.");
+		return -EINVAL;
+	}
+	data_len = hex2bin(arg_data, arg_data_len, data, sizeof(data));
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:999]:
```suggestion
	}
	
	data_len = hex2bin(arg_data, arg_data_len, data, sizeof(data));
```

File: subsys/bluetooth/shell/gatt.c [hunk 11]
```diff
@@ -946,43 +950,79 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
 
 static int cmd_notify(const struct shell *sh, size_t argc, char *argv[])
 {
-	struct bt_gatt_notify_params params;
-	uint8_t data = 0;
+	const struct bt_gatt_attr *attr;
+	int err;
+	size_t data_len;
+	unsigned long handle;
+	static char data[CHAR_SIZE_MAX];
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:978]: I am wondering if we could reuse the `gatt_write_buf`, as you cannot (currently anyhow) perform both the write and notify command in parallel, right?

**Review by Thalley — CHANGES_REQUESTED** [review]
A few code style comments and a suggestion to reuse existing buffer

File: subsys/bluetooth/shell/gatt.c [hunk 12]
```diff
@@ -946,43 +950,81 @@ static int cmd_unregister_test_svc(const struct shell *sh,
...
```
Comment (Thalley) [inline subsys/bluetooth/shell/gatt.c:978]: Also consider using `BT_ATT_MAX_ATTRIBUTE_LEN` instead of `CHAR_SIZE_MAX`

**Review by Thalley — DISMISSED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by theob-pro — APPROVED** [review]

**Pull request merged**